src/utils.c

```diff
@@ -110,30 +117,208 @@ SEXP colnamesInt(SEXP x, SEXP cols, SEXP check_dups) {
+    inverse = true;
+    expr = CADR(expr);
+  }
+  // non-evaluated case: V3:V2
+  // single symbol V2 may or may not be evaluated, see below
+  if (isLanguage(expr) && CAR(expr)==sym_colon) {  // 3:2, V3:V2, min(V3):min(V2)
+    //Rprintf("isLanguage(expr) && CAR(expr)==':'\n");
+    SEXP lhs = CADR(expr), rhs = CADDR(expr);
+    if (isSymbol(lhs) && isSymbol(rhs)) { // V3:V2
+      //Rprintf("isSymbol(colon_lhs) && isSymbol(colon_rhs)\n");
+      if (inverse)
+        error("column range selection by col1:col2 is not supported with combination of negation via '!' or '-'");
+      lhs = colnamesInt(x, ScalarString(PRINTNAME(lhs)), ScalarLogical(false), ScalarLogical(false)); // may raise error if lhs column does not exists
+      rhs = colnamesInt(x, ScalarString(PRINTNAME(rhs)), ScalarLogical(false), ScalarLogical(false));
+      if (!isInteger(lhs) || !isInteger(rhs) || length(lhs)!=1 || length(rhs)!=1 || LOGICAL(lhs)[0]==NA_LOGICAL || LOGICAL(rhs)[0]==NA_LOGICAL)
+        error(_("internal error: LHS and RHS of `:` call should be integer non-NA scalars already")); // # nocov
+      UNPROTECT(protecti);
+      return range_int(INTEGER(lhs)[0], INTEGER(rhs)[0]);
+    } else { // evaluates later on: 3:2, f(V3):f(V2)
+      //Rprintf("!isSymbol(colon_lhs) || !isSymbol(colon_rhs)\n");
+    }
+  }
+  SEXP cols = R_NilValue, value = R_NilValue;
+  if (isLanguage(expr) && CAR(expr)==sym_patterns) {
+    //Rprintf("isLanguage(expr) && CAR(expr)=='patterns'\n");
+    SEXP xnames = getAttrib(x, R_NamesSymbol);
+    if (isNull(xnames))
+      error(_("'x' argument data.table has no names"));
+    // .SDcols = Reduce(intersect, do_patterns(colsub, names_x))
+    SEXP sym_quote = install("quote");
+    SEXP sym_do_patterns = install("do_patterns");
+    SEXP lpat = PROTECT(eval(LCONS(sym_do_patterns, LCONS(LCONS(sym_quote, LCONS(expr, R_NilValue)), LCONS(xnames, R_NilValue))), rho)); protecti++;
+    SEXP sym_Reduce = install("Reduce");
+    SEXP sym_intersect = install("intersect");
+    cols = PROTECT(eval(LCONS(sym_Reduce, LCONS(sym_intersect, LCONS(lpat, R_NilValue))), rho)); protecti++;
+  }
+  // single symbol V2 might be also a function, to check that we need to evaluate, thus first we check if "V2" is existing column name, if not then we evaluate to see if it is a function.
+  // adding support of 'with' argument here may improve control, when with=FALSE then cols=eval(expr, rho) could be made straight away
+  if (isSymbol(expr)) { // V1, is.numeric
+    if (!isNull(cols))
+      error("internal error: cols should be still NULL at that point, did CAR(expr)=='patterns' branch kicked in? then isSymbol(expr) branch should not be reached"); // # nocov
+    SEXP xnames = getAttrib(x, R_NamesSymbol);
+    if (isNull(xnames))
+      error(_("'x' argument data.table has no names"));
+    cols = PROTECT(ScalarString(PRINTNAME(expr))); protecti++; // "V1", "is.numeric"
+    if (INTEGER(chmatch(cols, xnames, 0))[0]==0)
```

Comment:
You could go straight into the ```//evaluate expr``` right here instead of another ```if``` statement. I did something similar in R. 
```
# if colsub is also a name in the data.table, we prioritize it over anything in the environment
if (exists(as.character(colsub), where = x)) {
  cols = as.character(colsub) 
} else {
  cols = eval(colsub, parent.frame(2L), parent.frame(2L))
}
```

Reply:
I could but I will need to eval later on anyway for other cases, so it reduces lines of code to eval once.

Reply:
I put comments below, feel free to unresolved this thread if you see something more to do about that.